metrics/amplitude.js

```diff
@@ -0,0 +1,332 @@
+      let serviceName, clientId;
+
+      const { service } = data;
+      if (service && service !== 'content-server') {
+        if (service === 'sync') {
+          serviceName = service
+        } else {
+          serviceName = services[service] || 'undefined_oauth'
+          clientId = service
+        }
+      }
+
+      return { serviceName, clientId }
+    }
+
+    function mapUserProperties (eventGroup, eventCategory, data) {
+      return Object.assign(
+        pruneUnsetValues({
+          entrypoint: data.entrypoint,
+          flow_id: data.flowId,
+          ua_browser: data.browser,
+          ua_version: data.browserVersion,
+          utm_campaign: data.utm_campaign,
+          utm_content: data.utm_content,
+          utm_medium: data.utm_medium,
+          utm_source: data.utm_source,
+          utm_term: data.utm_term
+        }),
+        mapAppendProperties(data),
+        mapSyncDevices(data),
+        mapNewsletterState(eventCategory, data)
+      );
+    }
+
+    function mapAppendProperties (data) {
+      const servicesUsed = mapServicesUsed(data);
+      const experiments = mapExperiments(data);
+
+      if (servicesUsed || experiments) {
+        return {
+          '$append': Object.assign({}, servicesUsed, experiments)
+        };
+      }
+    }
+
+    function mapServicesUsed (data) {
+      const { serviceName } = getServiceNameAndClientId(data);
+
+      if (serviceName) {
+        return {
+          fxa_services_used: serviceName
+        };
+      }
+    }
+  }
+};
+
+function pruneUnsetValues (data) {
+  const result = {};
+
+  Object.keys(data).forEach(key => {
+    const value = data[key];
+
+    if (value || value === false) {
+      result[key] = value;
+    }
+  })
+
+  return result;
+}
+
+function mapExperiments (data) {
+  const { experiments } = data;
+
+  if (Array.isArray(experiments) && experiments.length > 0) {
+    return {
+      experiments: experiments.map(e => `${toSnakeCase(e.choice)}_${toSnakeCase(e.group)}`)
+    };
+  }
+}
+
+function toSnakeCase (string) {
+  return string.replace(/([a-z])([A-Z])/g, (s, c1, c2) => `${c1}_${c2.toLowerCase()}`)
+    .replace(/([A-Z])/g, c => c.toLowerCase())
+    .replace(/\./g, '_')
+    .replace(/-/g, '_');
+}
+
+function mapSyncDevices (data) {
+  const { devices } = data;
+
+  if (Array.isArray(devices)) {
+    return {
+      sync_device_count: devices.length,
+      sync_active_devices_day: countDevices(devices, DAY),
+      sync_active_devices_week: countDevices(devices, WEEK),
+      sync_active_devices_month: countDevices(devices, FOUR_WEEKS)
+    };
+  }
+}
+
+function countDevices (devices, period) {
+  return devices.filter(device => device.lastAccessTime >= Date.now() - period).length
+}
+
+function mapNewsletterState (eventCategory, data) {
+  let newsletter_state = NEWSLETTER_STATES[eventCategory];
+
+  if (! newsletter_state) {
+    const { marketingOptIn } = data;
+
+    if (marketingOptIn === true || marketingOptIn === false) {
```

Comment:
I have to admit I was confused by this, are you just trying to filter out non-boolean types?

Reply:
Yep, my intent is to make sure `undefined` does not produce the same effect as `false`.